server/channels/app/platform/web_hub.go

```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
 		return
 	}
 
+	// Remove the wc from i.byUserId
 	// get the conn slice.
 	userConnections := i.byUserId[wc.UserId]
 	// get the last connection.
 	last := userConnections[len(userConnections)-1]
-	// set the slot that we are trying to remove to be the last connection.
+	// https://go.dev/wiki/SliceTricks#delete-without-preserving-order
 	userConnections[userConnIndex] = last
-	// remove the last connection pointer from slice.
 	userConnections[len(userConnections)-1] = nil
-	// remove the last connection from the slice.
 	i.byUserId[wc.UserId] = userConnections[:len(userConnections)-1]
 	// set the index of the connection that was moved to the new index.
 	i.byConnection[last] = userConnIndex
 
+	connectionID := wc.GetConnectionID()
+	// Remove webconns from i.byChannelID
+	// This has O(n) complexity. We are trading off speed while removing
+	// a connection, to improve broadcasting a message.
+	for chID, webConns := range i.byChannelID {
+		// https://go.dev/wiki/SliceTricks#filtering-without-allocating
+		filtered := webConns[:0]
+		for _, conn := range webConns {
+			if conn.GetConnectionID() != connectionID {
+				filtered = append(filtered, conn)
+			}
+		}
+		for i := len(filtered); i < len(webConns); i++ {
+			webConns[i] = nil
+		}
+		i.byChannelID[chID] = filtered
+	}
+
 	delete(i.byConnection, wc)
-	delete(i.byConnectionId, wc.GetConnectionID())
+	delete(i.byConnectionId, connectionID)
+}
+
+func (i *hubConnectionIndex) InvalidateCMCacheForUser(userID string) {
+	// Clear out all user entries which belong to channels.
+	for chID, webConns := range i.byChannelID {
+		// https://go.dev/wiki/SliceTricks#filtering-without-allocating
+		filtered := webConns[:0]
+		for _, conn := range webConns {
+			if conn.UserId != userID {
+				filtered = append(filtered, conn)
+			}
+		}
+		for i := len(filtered); i < len(webConns); i++ {
+			webConns[i] = nil
+		}
+		i.byChannelID[chID] = filtered
+	}
+
+	// re-populate the cache
+	cm, err := i.getChannelMembersForUser(userID)
+	if err != nil {
+		panic(fmt.Errorf("error while getting GetAllChannelMembersForUser: %v", err))
+	}
+	for chID := range cm {
+		i.byChannelID[chID] = append(i.byChannelID[chID], i.ForUser(userID)...)
+	}
+}
+
+const numRetries = 2
+
+// getChannelMembersForUser is a wrapper around the store method which retries
+// for numRetries times. After that we panic, to fail loudly because it is even
+// worse to continue without the channel membership data.
+// Note: the hub has internal recovery mechanism to auto-restart in case of a panic.
+// So only one hub out of numCPU hubs will be affected.
```

Comment:
I'd prefer if we prevented the specific user from connecting rather than crashing the hub along with all connections (from multiple users) in them.

Reply:
That will be difficult at the hub level because this data is used to control broadcast data for all users in the hub. The per-user control was possible when this was at the web_conn level.